art/attacks/evasion/fast_gradient.py

```diff
@@ -431,20 +431,19 @@ def _apply_norm(norm, grad, object_type=False):
             ).any():
                 logger.info("The loss gradient array contains at least one positive or negative infinity.")
 
+            grad_2d = grad.reshape(1 if object_type else len(grad), -1)
             if norm in [np.inf, "inf"]:
-                grad = np.sign(grad)
+                grad_2d = np.ones_like(grad_2d)
             elif norm == 1:
-                if not object_type:
-                    ind = tuple(range(1, len(x.shape)))
-                else:
-                    ind = None
-                grad = grad / (np.sum(np.abs(grad), axis=ind, keepdims=True) + tol)
-            elif norm == 2:
-                if not object_type:
-                    ind = tuple(range(1, len(x.shape)))
-                else:
-                    ind = None
-                grad = grad / (np.sqrt(np.sum(np.square(grad), axis=ind, keepdims=True)) + tol)
+                i_max = np.argmax(np.abs(grad_2d), axis=1)
+                grad_2d = np.zeros_like(grad_2d)
+                grad_2d[range(len(grad_2d)), i_max] = 1
+            elif norm > 1:
+                conjugate = norm / (norm - 1)
+                q_norm = np.linalg.norm(grad_2d, ord=conjugate, axis=1, keepdims=True)
+                with np.errstate(divide="ignore"):
+                    grad_2d = (np.abs(grad_2d) * np.where(q_norm, 1 / q_norm, 0)) ** (conjugate - 1)
+            grad = grad_2d.reshape(grad.shape) * np.sign(grad)
             return grad
 
         # Add momentum
```

Comment:
@eliegoudout I think the failing test is cause by 'norm=1' in line 451 of fast_gradient.py. What do you think, that line should be `grad = _apply_norm(norm=self.norm, grad=grad)`?

Reply:
If yes, this was a bug that already existed before this PR and got discovered by your changes.

Reply:
Oooh you might be right, nice catch! Indeed, `_apply_norm` should be called with the right `norm` parameter (that is the $p$ of $L^p$, instead of $1$). This also explain perfectly the observed smaller $x-x_{\text{adv}}$, since with this PR, FGM($p=1$) corresponds to only changing the most important input feature (where before, the approach changesd many pixels).

Reply:
@eliegoudout Wow, that's great. Thank you for checking the paper, I think your explanation is correct. If not too late I'm wondering if we should add a comment to line 451 to mention that this is a choice by the original paper?

Reply:
Yes, I will add a comment, it is a good idea to clarify things I think 👍

Reply:
By the way, I think there might be a confusion here (before PR change)
```python
# Add momentum
if decay is not None and momentum is not None:
    grad = _apply_norm(norm=1, grad=grad)
    grad = decay * momentum + grad
    momentum += grad
```
It seems to me that this effectively gives (with $m=\text{momentum}$):
$$m_t = (1+\text{decay})\times m_{t-1} + \text{grad}_t/\Vert\text{grad}_t\Vert_1$$
instead of the paper's

$$m_t = \text{decay}\times m_{t-1} + \text{grad}_t/\Vert\text{grad}_t\Vert_1.$$
Do you agree @beat-buesser?

Furthermore, I think `grad` and `_apply_norm` are poor naming conventions and should be changed. But if we decide to improve them, it should be done later I think.

Reply:
> One more note: I think that the design choice of modifying `momentum` only by side effect is pretty confusing. I'll probably leave it like that though, to avoid adding unnecessary refactoring.

Well... as it happens, I just verified, and for reasons I don't know, the in-place modification of momentum doesn't always seem to work. For example, I tried changing the decay factor and got the exact same results: passing tests *both* following cases...
```python
attack = MomentumIterativeMethod(classifier, eps=0.3, eps_step=0.1, decay=1.0, max_iter=10)
x_train_mnist_adv = attack.generate(x=x_train_mnist, y=y_train_mnist)
attack2 = MomentumIterativeMethod(classifier, eps=0.3, eps_step=0.1, decay=123456, max_iter=10)
x_train_mnist_adv2 = attack.generate(x=x_train_mnist, y=y_train_mnist)
assert np.all(x_train_mnist_adv == x_train_mnist_adv2)  # PASSES WHEN IT SHOULD NOT
```

I don't know if this is specific to the tensorflow framework or not, I don't really know how to control which framework pytest uses. In any case, there seems to be a genuine implementation bug regarding momentum.

Reply:
Okay, I come with some more understanding of the situation after having implemented my fix regarding the correct computation of the momentum.

For context, the test `test_images` from `test_momentum_iterative_method.py` is essentially as follows:
```python
attack = MomentumIterativeMethod(classifier, eps=0.3, eps_step=0.1, decay=1.0, max_iter=10)
x_train_mnist_adv = attack.generate(x=x_train_mnist)

assert np.mean(np.abs(x_train_mnist - x_train_mnist_adv)) == pytest.approx(0.09437845647335052, abs=0.05)
assert np.max(np.abs(x_train_mnist - x_train_mnist_adv)) == pytest.approx(0.3)
```

and I was really surprised about the error tolerance `abs=0.5` from the first line (>50%!)...

After having updated the code to include my computation of the momentum, I reran the tests with different values for `decay` and frameworks numpy, tf and pytorch. Here are the (surprising) results for the mean of the differences obtained:
```python
# NumPy fixed
{None: 0.09437846,
 0: 0.09437846,
 0.5: 0.1246982,
 1.0: 0.12893921,
 1000: 0.12811291}

# PyTorch fixed
{None: 0.09406391,
 0: 0.09406391,
 0.5: 0.124101475,
 1.0: 0.12861145,
 1000:0.128003241}

# TensorFlow fixed
{None: 0.09437846,
 0: 0.09437846,
 0.5: 0.09437846,
 1.0: 0.09437846,
 1000:0.094378461}
```

My conclusions are as follow:
1. The in-place modification of `momentum` doesn't work, only for TensorFlow (maybe someone knows why?),
2. The person designing the test probably ran them with 2 different framework and concluded, from the size of the difference in results, that it was appropriate to have `abs=0.5`.
3. It seems clear that in fact, the tensorflow version of the momentum iterative method was not working at all (and still isn't!). As I said, I think this might involve refacto, so is left for future PR.
4. In the mean time, we have to statute regarding the tests. I think we should replace the current check with the following, and skip the test for tensorflow atm
```python
# Proposed new check
assert np.mean(np.abs(x_train_mnist - x_train_mnist_adv)) == pytest.approx(0.1288, abs=5e-4)
```

Reply:
Hi @eliegoudout That's a great update, let me take a closer look.

Reply:
Hi @eliegoudout Thank you very much for your detailed analysis. I agree with your interpretation. The equation for momentum is not implemented correctly. I think removing the wrong inlace addition might also fix the issue with TensorFlow. I like the new test tolerance. We could consider defining it as a relative tolerance, which would make it more interpretable.

Reply:
Yes, I agree, let's fix the momentum issue in a different issue and pull request.